scipy/stats/_survival.py

```diff
@@ -374,3 +384,122 @@ def _ecdf_right_censored(sample):
     cdf = 1 - sf
     return t, cdf, sf, n, d
+
+
+@dataclass
+class LogRankResult:
+    statistic: np.ndarray
+    pvalue: np.ndarray
+
+
+def _at_risk(times: np.ndarray, all_times: np.ndarray) -> np.ndarray:
+    """Times need to be sorted."""
+    at_risk = []
+    n_times = len(times)
+    for x in all_times:
+        idx_ = bisect.bisect_left(times, x)  # type: ignore[call-overload]
+        at_risk.append(n_times-idx_)
+    return np.array(at_risk)
+
+
+def _iv_log_rank(
+    sample: CensoredData, control: CensoredData,
+) -> tuple[CensoredData, CensoredData]:
+    if not (isinstance(sample, CensoredData) and
+            isinstance(control, CensoredData)):
+        raise ValueError(
+            ""
+        )
+
+    return sample, control
+
+
+def log_rank(sample: CensoredData, control: CensoredData) -> LogRankResult:
+    """Log Rank test.
+
+    Parameters
+    ----------
+    sample, control : CensoredData
+        Sample and control data to compare based on their survival functions.
+        The order does not matter, `sample` and `control` can be interverted.
+
+    Returns
+    -------
+    res : LogRankResult
+        An object containing attributes:
+
+        statistic : float ndarray
+            The computed statistic of the test.
+        pvalue : float ndarray
+            The computed p-value of the test.
+
+    References
+    ----------
+    .. [1] ...
+
+    Examples
+    --------
+
+    >>> from scipy import stats
+    >>> sample = stats.CensoredData(
+    ...     uncensored=[ 8., 12., 26., 14., 21., 27.],
+    ...     right=[ 8., 32., 20., 40.]
+    ... )
+    >>> control = stats.CensoredData(
+    ...     uncensored=[33., 28., 41.],
+    ...     right=[48., 48., 25., 37., 48., 25., 43.]
+    ... )
+
+    >>> import numpy as np
+    >>> import matplotlib.pyplot as plt
+    >>> ax = plt.subplot()
+    >>> ecdf_sample = stats.ecdf(sample)
+    >>> ax.step(np.insert(ecdf_sample.x, 0, 0), np.insert(ecdf_sample.sf.points, 0, 1),
+    ...         where='post', label='sample')
+    >>> ecdf_control = stats.ecdf(control)
+    >>> ax.step(np.insert(ecdf_control.x, 0, 0), np.insert(ecdf_control.sf.points, 0, 1),
+    ...         ls='-.', where='post', label='control')
+    >>> ax.set_xlabel('Time (months)')
+    >>> ax.set_ylabel('Empirical CDF')
+    >>> plt.legend()
+    >>> plt.show()
+
+    >>> res = stats.log_rank(sample=sample, control=control)
+    >>> res.statistic
+    6.148087536256203
+    >>> res.pvalue
+    0.013155428547469983
+
+    """
+    sample, control = _iv_log_rank(sample=sample, control=control)
+
+    sample_tot = CensoredData(
+        uncensored=np.concatenate((sample._uncensored, control._uncensored)),
+        right=np.concatenate((sample._right, control._right))
+    )
+
+    res = ecdf(sample_tot)
+    times_sample_tot = res.x[res.sf._d.astype(bool)]
+    at_risk_sample_tot = res.sf._n[res.sf._d.astype(bool)]
+    _, dead_per_times = np.unique(sample_tot._uncensored, return_counts=True)
```

Comment:
```suggestion
    i = res.sf._d.astype(bool)  # events were observed (not censored)
    times_xy= res.x[i]
    at_risk_xy = res.sf._n[i]
    deaths_xy = res.sf._d[i]
```
Right?

`tot` instead of `xy`, if you prefer, but I think the last variable should have the same suffix at the rest.